File: ioloolo/BOJ/4949.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

public class Main {
	BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
	BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

	List<String> texts = new LinkedList<>();

	void input() throws IOException {
		String line;
		while (!(line = reader.readLine()).equals(".")) {
			texts.add(line);
		}
	}

	void solve() throws IOException {
		for (String text : texts) {
			Stack<Character> stack = new Stack<>();

			for (char c : text.toCharArray()) {
				if (c == '(' || c == '[') stack.push(c);

				if (c == ')') {
					if (!stack.isEmpty() && stack.peek() == '(') {
						stack.pop();
					} else {
						stack.push('A');
						break;
					}
				}

				if (c == ']') {
					if (!stack.isEmpty() && stack.peek() == '[') {
						stack.pop();
					} else {
						stack.push('A');
						break;
					}
				}
			}

			writer.write(stack.isEmpty() ? "yes" : "no");
			writer.newLine();
		}
	}

	public static void main(String[] args) throws IOException {
		Main main = new Main();

		main.input();
		main.solve();

		main.reader.close();
		main.writer.close();
	}
}

File: wlwl1011/BOJ/LG/stack/4949.py
import sys
input = lambda : sys.stdin.readline().rstrip('\r\n')

s = input()
if s == '.':
    print('no')
else:    

    while s != '.':
        stack = []
        flag = True
        for i in range(len(s)-1): #맨끝은 온점으로 끝
           
            if s[i] == '(' or s[i] == '[':
                stack.append(s[i])
            elif s[i] == ')':
                if stack and stack[-1] == '(':
                    stack.pop()
                
                else:
                    print('no')
                    flag = False
                    break
            elif s[i] == ']':
                if stack and stack[-1] == '[':
                    stack.pop()    
                else:
                    flag = False
                    print('no')
                    break    
        if stack and flag:
            print('no')
        elif flag :
            print('yes')    
        s = input()    
   

